packages/oauth/src/install-provider.ts

```diff
@@ -0,0 +1,654 @@
+            v2Installation.enterpriseUrl = authResult.url;
+          }
+
+          // Token Rotation is Enabled
+          if (v2Resp.authed_user.refresh_token !== undefined && v2Resp.authed_user.expires_in !== undefined) {
+            v2Installation.user.refreshToken = v2Resp.authed_user.refresh_token;
+            v2Installation.user.expiresAt = currentUTC + v2Resp.authed_user.expires_in; // utc, seconds
+          }
+        }
+
+        resp = v2Resp;
+        installation = v2Installation;
+      }
+
+      if (resp.incoming_webhook !== undefined) {
+        installation.incomingWebhook = {
+          url: resp.incoming_webhook.url,
+          channel: resp.incoming_webhook.channel,
+          channelId: resp.incoming_webhook.channel_id,
+          configurationUrl: resp.incoming_webhook.configuration_url,
+        };
+      }
+      if (installOptions && installOptions.metadata !== undefined) {
+        // Pass the metadata in state parameter if exists.
+        // Developers can use the value for additional/custom data associated with the installation.
+        installation.metadata = installOptions.metadata;
+      }
+      // End: Build the installation object
+
+      // Save installation object to installation store
+      if (installation.isEnterpriseInstall) {
+        await this.installationStore.storeInstallation(installation as OrgInstallation, this.logger);
+      } else {
+        await this.installationStore.storeInstallation(installation as Installation<'v1' | 'v2', false>, this.logger);
+      }
+
+      // Call the success callback
+      if (options !== undefined && options.success !== undefined) {
+        this.logger.debug('calling passed in options.success');
+        options.success(installation, installOptions, req, res);
+      } else {
+        this.logger.debug('run built-in success function');
+        defaultCallbackSuccess(installation, installOptions, req, res);
+      }
+    } catch (error: any) {
+      this.logger.error(error);
+
+      if (!installOptions) {
+        // To make the `installOptions` type compatible with `CallbackOptions#failure` signature
+        const emptyInstallOptions: InstallURLOptions = { scopes: [] };
+        // eslint-disable-next-line no-param-reassign
+        installOptions = emptyInstallOptions;
+      }
+
+      // Call the failure callback
+      if (options !== undefined && options.failure !== undefined) {
+        this.logger.debug('calling passed in options.failure');
+        options.failure(error, installOptions, req, res);
+      } else {
+        this.logger.debug('run built-in failure function');
+        defaultCallbackFailure(error, installOptions, req, res);
+      }
+    }
+  }
+}
+
+// This is intentionally structurally identical to AuthorizeResult from App
+// It is redefined so that this class remains loosely coupled to the rest
+// of Bolt.
+export interface AuthorizeResult {
```

Comment:
We may want to have dedicated files for these interface. However, I didn't do so at this moment because these are used only in `InstallProvider`. If we want to make it more granular, we can do so in future changes.

Reply:
Also, these are exported in `index.ts` as with before

Reply:
Revisited this and I came to think that `AuthorizeResult` should be `authorize-result.ts` as external code relies on this interface.